### ISSUE TITLE
Investigate memory quota issue on data import

### ISSUE THREAD

**Issue by alexsoble** [studentinsights/studentinsights#11]
`Error R14 (Memory quota exceeded)`

**Comment by alexsoble** [comment]
https://github.com/codeforamerica/somerville-teacher-tool/issues/445

**Comment by alexsoble** [comment]
Baseline:

+ We are running tiny infrastructure: 1 Heroku hobby-size web dyno.
+ Ships with 512MB RAM.

**Comment by alexsoble** [comment]
This is starting to become a more relevant issue. Import task crashed in staging on attendance import as the quota errors escalated. Will see if I can replicate in production.

**Comment by alexsoble** [comment]
Verified that the production import task didn't make it all the way through. 

Unfortunately it swallowed the error as well. 😕  

I wonder if this has something to do with the fact that #9 made `students_export.txt` much larger?

**Comment by alexsoble** [comment]
Heroku vs. local machine baseline: process takes 825 MB of memory running locally on my Mac.

**Comment by alexsoble** [comment]
Grows to 1.7 GB when handling `attendance_export.txt`.

**Comment by alexsoble** [comment]
A couple of different strategies here: 

1. Dive into the internals of the import process and see trade-offs we can make to save on memory. 

2. Break the import process into two different parts: one that does bulk import at when the database is empty (currently exists only for `attendance_export.txt`), and another, less memory-intensive task that updates only recent events.

Will probably take a mix of both to fix.  :0)

**Comment by alexsoble** [comment]
Profiled importing the smallest flat file (`educators_export.txt`) in staging environment to get a sense of what's eating up our memory. 

```
 Total retained:  65961 bytes (799 objects)
 
 allocated memory by gem
 -----------------------------------
    4321302  ruby-2.3.0/lib
    3349411  activerecord-4.1.14.1
    2768946  app/app
    1458990  net-ssh-3.0.2
    1016606  activesupport-4.1.14.1
     706264  arel-5.0.1.20140414130214
     644317  activemodel-4.1.14.1
      12185  net-sftp-2.1.2
       8640  devise-3.5.6
       2144  thread_safe-0.3.5
 
 allocated memory by file
 -----------------------------------
    4312206  /app/vendor/ruby-2.3.0/lib/ruby/2.3.0/csv.rb
    1582942  /app/app/reports/progress_bar.rb
     684152  /app/app/importers/data_transformers/csv_row_cleaner.rb
     546685  /app/vendor/bundle/ruby/2.3.0/gems/activemodel-4.1.14.1/lib/active_model/attribute_methods.rb
     528552  /app/vendor/bundle/ruby/2.3.0/gems/activerecord-4.1.14.1/lib/active_record/relation/query_methods.rb
     455106  /app/vendor/bundle/ruby/2.3.0/gems/net-ssh-3.0.2/lib/net/ssh/transport/packet_stream.rb
     453887  /app/vendor/bundle/ruby/2.3.0/gems/arel-5.0.1.20140414130214/lib/arel/visitors/to_sql.rb
     418188  /app/vendor/bundle/ruby/2.3.0/gems/net-ssh-3.0.2/lib/net/ssh/buffer.rb
     391712  /app/vendor/bundle/ruby/2.3.0/gems/activerecord-4.1.14.1/lib/active_record/relation.rb
     339303  /app/vendor/bundle/ruby/2.3.0/gems/activerecord-4.1.14.1/lib/active_record/connection_adapters/postgresql/database_statements.rb
     257872  /app/app/importers/data_transformers/csv_transformer.rb
```

**Comment by alexsoble** [comment]
Removing unnecessary string objects from `ProgressBar` via 628e6e2149cb3d6d2992ef68e9ea038dcc6bd648 reduces memory allocated to `ProgressBar` by 42%:

```
 allocated memory by file
 -----------------------------------
    4312206  /app/vendor/ruby-2.3.0/lib/ruby/2.3.0/csv.rb
     947818  /app/vendor/bundle/ruby/2.3.0/gems/net-ssh-3.0.2/lib/net/ssh/transport/packet_stream.rb
     684152  /app/app/importers/data_transformers/csv_row_cleaner.rb
     669913  /app/app/reports/progress_bar.rb
     546685  /app/vendor/bundle/ruby/2.3.0/gems/activemodel-4.1.14.1/lib/active_model/attribute_methods.rb
     528552  /app/vendor/bundle/ruby/2.3.0/gems/activerecord-4.1.14.1/lib/active_record/relation/query_methods.rb
```

**Comment by alexsoble** [comment]
Looking for more refactoring targets...

`ruby/2.3.0/csv.rb:243` is an interesting one, looks like it has to do with handling extra headers or fields when CSV parsing:

```
# https://github.com/ruby/ruby/blob/trunk/lib/csv.rb#L241 =>

241      # handle extra headers or fields
242      @row = if headers.size >= fields.size
243        headers.zip(fields)
244      else
245        fields.zip(headers).map { |pair| pair.reverse! }
246      end
``` 

Wonder if we have lots of mismatched header/field sizes?

**Comment by alexsoble** [comment]
Looks like we can allocate almost 50% less memory in `ruby/2.3.0/csv.rb` by cutting out duplicate CSV parsing.

Here's the profile for allocated memory by file, same task as above, after implementing adb63b90515eeb3b8992fe99d679b491994d1ffa:

```
 allocated memory by file
 -----------------------------------
    2283524  /app/vendor/ruby-2.3.0/lib/ruby/2.3.0/csv.rb
     684152  /app/app/importers/data_transformers/csv_row_cleaner.rb
     669913  /app/app/reports/progress_bar.rb
     546685  /app/vendor/bundle/ruby/2.3.0/gems/activemodel-4.1.14.1/lib/active_model/attribute_methods.rb
     528552  /app/vendor/bundle/ruby/2.3.0/gems/activerecord-4.1.14.1/lib/active_record/relation/query_methods.rb
     474066  /app/vendor/bundle/ruby/2.3.0/gems/net-ssh-3.0.2/lib/net/ssh/transport/packet_stream.rb
     453887  /app/vendor/bundle/ruby/2.3.0/gems/arel-5.0.1.20140414130214/lib/arel/visitors/to_sql.rb
     391712  /app/vendor/bundle/ruby/2.3.0/gems/activerecord-4.1.14.1/lib/active_record/relation.rb
     383084  /app/vendor/bundle/ruby/2.3.0/gems/net-ssh-3.0.2/lib/net/ssh/buffer.rb
     339303  /app/vendor/bundle/ruby/2.3.0/gems/activerecord-4.1.14.1/lib/active_record/connection_adapters/postgresql/database_statements.rb
```

😊

**Comment by kevinrobinson** [comment]
@alexsoble you are awesome sir

**Comment by alexsoble** [comment]
Heh I am just fixing a memory mess of my own making!

**Comment by alexsoble** [comment]
Profiling results from 13b18c23c5044eeef26e26ae5d2561f1507fdd7f: 

```
 allocated memory by file
 -----------------------------------
    2283524  /app/vendor/ruby-2.3.0/lib/ruby/2.3.0/csv.rb
     753306  /app/vendor/bundle/ruby/2.3.0/gems/net-ssh-3.0.2/lib/net/ssh/transport/packet_stream.rb
     684152  /app/app/importers/data_transformers/csv_row_cleaner.rb
     546685  /app/vendor/bundle/ruby/2.3.0/gems/activemodel-4.1.14.1/lib/active_model/attribute_methods.rb
     528552  /app/vendor/bundle/ruby/2.3.0/gems/activerecord-4.1.14.1/lib/active_record/relation/query_methods.rb
     453887  /app/vendor/bundle/ruby/2.3.0/gems/arel-5.0.1.20140414130214/lib/arel/visitors/to_sql.rb
     393877  /app/vendor/bundle/ruby/2.3.0/gems/net-ssh-3.0.2/lib/net/ssh/buffer.rb
     391712  /app/vendor/bundle/ruby/2.3.0/gems/activerecord-4.1.14.1/lib/active_record/relation.rb
     339303  /app/vendor/bundle/ruby/2.3.0/gems/activerecord-4.1.14.1/lib/active_record/connection_adapters/postgresql/database_statements.rb
     257912  /app/app/importers/data_transformers/csv_transformer.rb
     209968  /app/vendor/bundle/ruby/2.3.0/gems/net-ssh-3.0.2/lib/net/ssh/transport/state.rb
```

**Comment by alexsoble** [comment]
Some good news: now that the CSV duplication issue is fixed and `ProgressBar` turned off in production, the entire import task can run in staging without triggering `R15 - Memory quota vastly exceeded` and being killed by `SIGTERM`, handling the much larger `students_export.txt` file from #9.

The task still causes Heroku to throw `Error R14 (Memory quota exceeded)`, which isn't great, but won't kill the process. A future refactor could cut down memory usage much farther, in particular by reading the CSVs line-by-line instead of parsing the whole file at once. Closing this issue for now since we've set up a profiling baseline, made all the easy fixes, and fixed the critical production errors.
